contracts/VAIController.sol

```diff
@@ -93,7 +104,14 @@ contract VAIController is VAIControllerStorageG2, VAIControllerErrorReporter, Ex
                 return fail(Error.REJECTION, FailureInfo.VAI_MINT_REJECTION);
             }
 
-            (vars.mathErr, vars.accountMintVAINew) = addUInt(ComptrollerImplInterface(address(comptroller)).mintedVAIs(minter), mintVAIAmount);
+            // Calculate the minted balance based on interest index
+            uint totalMintedVAI = ComptrollerImplInterface(address(comptroller)).mintedVAIs(minter);
+
+            if (totalMintedVAI > 0) {
+                totalMintedVAI = getVAIRepayAmount(minter);
+            }  
+
+            (vars.mathErr, vars.accountMintVAINew) = addUInt(totalMintedVAI, mintVAIAmount);
```

Comment:
So here we put interest into the principal which seems ok, and this is exactly what we do for VTokens. However, later we treat interest and principal separately to compute the percentages to burn and to put to treasury. This means that the following two sequences of actions will lead to different results:

```
Sequence A:
1. mintVAI(100 VAI)
2. wait for 1 year
3. mint(0 VAI)
4. repayVAI(MAX_UINT256)
```

```
Sequence B:
1. mintVAI(100 VAI)
2. wait for 1 year
3. repayVAI(MAX_UINT256)
```

In sequence A the entire interest will be burned, while in sequence B the interest will go into treasury.

Reply:
Yes I think you understood it correctly. Should we introduce another variable to track user's interest? And transfer to treasury whatever is in that variable.

Reply:
> This looks like a right thing to do, but I'm not sure what formulas we should use in this case. Does this look right?
> 
> During minting we save interest to a separate variable `pastInterest` (mapping from address to uint): pastInterest=mintedVAI∗(globalIndex/minterIndex) mintedVAI+=mintAmount minterIndex=globalIndex
> 
> Then to compute the owed amount and the proportion that would go to treasury/burn we'd use: owedAmount=(mintedVAI+pastInterest)∗(globalIndex/minterIndex) principal=mintedVAI (goes to burn) totalInterest=owedAmount−principal (goes to treasury)
> 
> What would be the formula for partial repayment? Will we subtract anything from pastInterest?

Yes I think in case of partial repayment we don't need to subtract anything from pastInterest.

Reply:
> Other options should be possible as well. Do we really need to track "treasury amount" per user? Maybe we can simplify the formulas by using global market-level math.

Do you have any formulas in mind? I cannot think of any.

Reply:
> Yes I think in case of partial repayment we don't need to subtract anything from pastInterest.

This makes the conditions kinda complex:
if repayAmount < mintedVAI, repay a part of the principal
else if repayAmount < mintedVAI + current interest, repay a part of the principal and a part of the current interest,
else if repayAmount < mintedVAI + past interest + current interest, **???**
otherwise, repay everything

I'd rather simplify this if possible, but I can't think of a formula from the top of my head. Something like track the total amount of minted VAIs (without interest) and total accumulated interest, compute a percentage based on these values, put this percentage to treasury and burn the rest? 🤔

Reply:
I think we can simplify the above by keeping mintedVAI as it is currently and just use totalInterest to track the amount of mintedVAI that's actually the interest amount. 

```
if repayAmount < mintedVAI
  mintedVAI = mintedVAI - repayAmount
else 
  mintedVAI = 0


if totalInterest > mintedVAI
  treasuryTransferAmount = totalInterest - mintedVAI
  totalInterest = mintedVAI
else 
  treasuryTransferAmount = totalInterest
  totalInterest = 0
```

Also we can include the currentInterest in the totalInterest before running the above formula.

Reply:
As discussed we will have `mintedVAI` as the total VAI minted by user and `totalInterest` representing interest accrued by the user

Reply:
@kkirka I have updated the implementation. Now we have a variable pastInterestOfUser tracking the amount of mintedVaiOfUser that represents the user's accrued interest. During repayment I use this formulas to update mintedVAI and pastInterest:

```
repayAmount = mintedVaiOfUser * (globalIndex / userMintIndex)
partOfRepayAmountUserIsRepaying = amountUserIsRepaying / repayAmount
totalCurrentInterestOfUser =repayAmount - mintedVaiOfUser +pastInterestOfUser 
partOfCurrentInterestOfUser =totalCurrentInterestOfUser *partOfRepayAmountUserIsRepaying 
treasuryAmount = treasuryAmount + partOfCurrentInterestOfUser
burn = (repayAmount - totalCurrentInterestOfUser) * partOfRepayAmountUserIsRepaying  
partOfPastInterestOfIser =pastInterestOfUser *partOfRepayAmountUserIsRepaying   
mintedVaiOfUser = mintedVaiOfUser - burn  -partOfPastInterestOfIser 
pastInterestOfUser=pastInterestOfUser -(pastInterestOfUser *partOfRepayAmountUserIsRepaying ) 
```